base_layer/blockchain/src/chain.rs

```diff
@@ -20,37 +20,125 @@
+            // block was processed fine and added to chain
+            Ok(_) => {
+                self.current_total_pow = new_block.header.pow;
+                // search for too old orphans
+                for (block_hash, block) in self.orphans.iter() {
+                    if block.header.height > MAX_ORPHAN_AGE {
+                        self.currently_used_orphans.push(block_hash.clone());
+                    }
+                }
+                for hash in &self.currently_used_orphans {
+                    // delete old orphans
+                    self.orphans.remove(hash);
+                }
+                self.currently_used_orphans = Vec::new();
+                Ok(())
+            },
+            // block seems valid, but its orphaned
+            Err(StateError::OrphanBlock) => {
+                if self.orphans.contains_key(&new_block.hash()[..]) {
+                    return Err(ChainError::StateProcessingError(StateError::DuplicateBlock));
+                };
+                let mut hash = [0; 32];
+                hash.copy_from_slice(&new_block.hash());
+                let pow = new_block.header.pow.clone();
+                self.orphans.insert(hash, new_block);
+                if pow > self.current_total_pow {
+                    // we have a potential re-org here
+                    let result = self.handle_re_org(&hash, 1);
+                    let result = if result.is_err() {
+                        self.block_chain_state.reset_chain_state()?;
+                        result
+                    } else {
+                        for hash in &self.currently_used_orphans {
+                            // delete old orphans
+                            self.orphans.remove(hash);
+                        }
+                        self.currently_used_orphans = Vec::new();
+                        self.block_chain_state
+                            .save_state()
+                            .map_err(ChainError::StateProcessingError)
+                    };
+                    self.currently_used_orphans = Vec::new();
+                    result
+                } else {
+                    Err(ChainError::StateProcessingError(StateError::OrphanBlock))
+                }
+            },
+            Err(e) => Err(ChainError::StateProcessingError(e)),
+        };
+        if result.is_err() {
+            self.block_chain_state.reset_chain_state()?;
+        }
+        result
     }
 
-    /// This block will validate the block and enforce the consensus rules on the block that dont require looking at
-    /// state, the all transactions have been signed, count up to zero commitments etc)
-    pub fn validate_new_block(&self, _new_block: &Block) -> Result<(), ChainError> {
-        Ok(())
+    /// Internal recursive function to handle re orgs
+    fn handle_re_org(&mut self, block_hash: &BlockHash, counter: usize) -> Result<(), ChainError> {
```

Comment:
This function didn't seem to work for the simple test case I tried:

A new block has a higher PoW than the tip, and was mined from a block three deep on the main chain